src/glob.zig

```diff
@@ -0,0 +1,3021 @@
+const Codepoint = CodepointIterator.Cursor.CodePointType;
+// const Codepoint = u32;
+const Cursor = CodepointIterator.Cursor;
+
+const CursorState = struct {
+    cursor: CodepointIterator.Cursor = .{},
+    /// The index in terms of codepoints
+    // cp_idx: usize,
+
+    fn init(iterator: *const CodepointIterator) CursorState {
+        var this_cursor: CodepointIterator.Cursor = .{};
+        _ = iterator.next(&this_cursor);
+        return .{
+            // .cp_idx = 0,
+            .cursor = this_cursor,
+        };
+    }
+
+    /// Return cursor pos of next codepoint without modifying the current.
+    ///
+    /// NOTE: If there is no next codepoint (cursor is at the last one), then
+    /// the returned cursor will have `c` as zero value and `i` will be >=
+    /// sourceBytes.len
+    fn peek(this: *const CursorState, iterator: *const CodepointIterator) CursorState {
+        var cpy = this.*;
+        // If outside of bounds
+        if (!iterator.next(&cpy.cursor)) {
+            // This will make `i >= sourceBytes.len`
+            cpy.cursor.i += cpy.cursor.width;
+            cpy.cursor.width = 1;
+            cpy.cursor.c = CodepointIterator.ZeroValue;
+        }
+        // cpy.cp_idx += 1;
+        return cpy;
+    }
+
+    fn bump(this: *CursorState, iterator: *const CodepointIterator) void {
+        if (!iterator.next(&this.cursor)) {
+            this.cursor.i += this.cursor.width;
+            this.cursor.width = 1;
+            this.cursor.c = CodepointIterator.ZeroValue;
+        }
+        // this.cp_idx += 1;
+    }
+
+    inline fn manualBumpAscii(this: *CursorState, i: u32, nextCp: Codepoint) void {
+        this.cursor.i += i;
+        this.cursor.c = nextCp;
+        this.cursor.width = 1;
+    }
+
+    inline fn manualPeekAscii(this: *CursorState, i: u32, nextCp: Codepoint) CursorState {
+        return .{
+            .cursor = CodepointIterator.Cursor{
+                .i = this.cursor.i + i,
+                .c = @truncate(nextCp),
+                .width = 1,
+            },
+        };
+    }
+};
+
+var dummyu32 = [_]u32{ 105, 102, 32, 121, 111, 117, 32, 115, 101, 101, 32, 116, 104, 105, 115, 44, 32, 122, 97, 99, 107, 32, 105, 115, 32, 97, 110, 32, 105, 100, 105, 111, 116 };
+
+pub const GlobWalker = struct {
+    pub const Result = Maybe(void);
+
+    arena: Arena = undefined,
+
+    /// not owned by this struct
+    pattern: []const u8 = "",
+
+    pattern_codepoints: []u32 = &dummyu32,
+    cp_len: u32 = 0,
+
+    /// If the pattern contains "./" or "../"
+    has_relative_components: bool = false,
+
+    patternComponents: ArrayList(Component) = .{},
+    matchedPaths: ArrayList(BunString) = .{},
+    i: u32 = 0,
+
+    dot: bool = false,
+    absolute: bool = false,
+    cwd: BunString = undefined,
+    follow_symlinks: bool = false,
+    error_on_broken_symlinks: bool = false,
+    only_files: bool = true,
+
+    root_cwd_slice: []const u8 = undefined,
```

Comment:
can this be `""` instead? `undefined` has caused strange issues many times
```suggestion
    root_cwd_slice: []const u8 = "",
```